src/ukify/ukify.py

```diff
@@ -1638,6 +1754,13 @@ def __call__(
         config_key='UKI/DeviceTreeAuto',
         config_push=ConfigItem.config_list_prepend,
     ),
+    ConfigItem(
+        '--hwids',
+        metavar='DIR',
+        type=Path,
+        help='Directory with HWID text files [.hwids section]',
+        config_key='UKI/HWIDs',
```

Comment:
there is no documentation for these files?

Reply:
yeah, where do they come from ? is there a repository with them available ?

Reply:
It's a concatenation of `fwupdtool hwids` and `Compatible: <compatible>`.
Examples are here: https://github.com/anonymix007/systemd-stub/tree/master/hwids

Reply:
What if the HW ID matches two of those (e.g. [sc8180x-lenovo-flex-5g-81xe.txt](https://github.com/anonymix007/systemd-stub/blob/master/hwids/sc8180x-lenovo-flex-5g-81xe.txt) and [sc8180x-lenovo-flex-5g-82ak.txt](https://github.com/anonymix007/systemd-stub/blob/master/hwids/sc8180x-lenovo-flex-5g-82ak.txt) )

Reply:
Both are `lenovo,flex-5g`, so the corresponding `.dtbauto` section will be used

Reply:
If I remember correctly, those two have different panels, so ideally they should use two distinct DTB files (not sure if the second one is properly supported upstream). Similarly there are different panel configurations used by some other laptops, which might require different DTB files to be selected (especially in the case of different types of panels).

Reply:
If those should use two distinct DTB files then those DTBs must have two distinct compatibles which doesn't seem to be the case as of now.
Stub can't do anything about it. In fact, dtbloader [has](https://github.com/TravMurav/dtbloader/blob/main/src/devices/lenovo_flex5g.c#L33) the same DTB path for both.

IMHO panel should not be in Devicetree in the first place. Surely EDID contains enough information to choose the proper driver, doesn't it?

Reply:
Not in case of the AMOLED panels, unfortunately. They frequently nead a bit of special treatment.

Reply:
Well, sure, but somehow AMOLED x86-64 laptops exist and don't require anything special. 
Surely the kernel can read EDID and choose the driver, can't it? I guess this technically relies on bootloader powering the screen on.

Reply:
@poettering what do you think of exposing libefi.a and all the necessary tools to build a EFI driver (in a separate package or something)? We might also introduce `src/boot/efi-drivers` with a few EFI drivers implementing DT Fixup protocol for commonly needed things (i.e. EDID for panels, [DPP](https://github.com/TravMurav/dtbloader/blob/dpp-mac/src/qcom.c) for BT/Wi-Fi MAC and so on).

Reply:
x86 laptops don't use DRM panel drivers to drive panels.
As far as I know, this can be brought down to different SKUs / Models in DMI. Thus I'm interested in the details on which HW IDs are used and in which order are they compared.

Reply:
Order is defined [here](https://github.com/systemd/systemd/blob/main/src/boot/chid.c#L103) and specific IDs are defined [here](https://github.com/systemd/systemd/blob/main/src/fundamental/chid-fundamental.c#L64) as an array of lists of strings to concatenate and hash.
So if you provide a set of HWID JSON files with different `compatible`s for laptops with different SKUs then different DTBs will be selected. 
Basically, always add JSON for the specific model/SKU and it should be handled properly.

Reply:
@anonymix007 do you have any plan/idea to add this functionality to non-uki boot flow ?

Reply:
@obbardc this implementation is based on [dtbloader](https://github.com/TravMurav/dtbloader) which can be used in a "non-uki boot flow".

Reply:
@anonymix007 right, but I was wondering if you had any plan/motivation to use the implementation already baked into sd-boot.

Reply:
It doesn't seem to be any easy task to integrate this into sd-boot and tbh it looks like a wrong place to integrate it too. 
If you really want this functionality, you may just use dtbloader or build UKI only with .dtbauto and .hwids and pretend that it's the kernel. I don't really see how this would benefit anyone or would even look like. Like where DTBs would come from? UKIs kind of make sense for this use case.